### PR TITLE
Support & use stable endpoints for MSC4151

### PR DESCRIPTION
**Pull request by turt2live**
https://github.com/matrix-org/matrix-spec-proposals/pull/4151 has finished FCP.

See https://github.com/element-hq/synapse/issues/17373 for unstable endpoint removal

### CODE REVIEW

File: synapse/rest/client/reporting.py [hunk 1]
```diff
@@ -155,8 +157,49 @@ async def on_POST(
...
+        return 200, {}
+
+
 def register_servlets(hs: "HomeServer", http_server: HttpServer) -> None:
     ReportEventRestServlet(hs).register(http_server)
+    ReportRoomRestServlet(hs).register(http_server)
 
     if hs.config.experimental.msc4151_enabled:
-        ReportRoomRestServlet(hs).register(http_server)
+        UnstableReportRoomRestServlet(hs).register(http_server)
```
Comment (anoadragon453) [inline synapse/rest/client/reporting.py:205]: Given we don't expect the functionality to differ between the stable and unstable versions of this endpoint, you can avoid the code duplication by setting `PATTERNS` from within the `__init__` method of `ReportRoomRestServlet` based on the experimental config value:

```py
class ReportRoomRestServlet(RestServlet): 
   # Note: Top-level PATTERNS definition removed
 
    def __init__(self, hs: "HomeServer"):
        super().__init__()
        self.hs = hs
        self.auth = hs.get_auth()
        self.clock = hs.get_clock()
        self.store = hs.get_datastores().main

        self.PATTERNS = client_patterns(
	        "/rooms/(?P<room_id>[^/]*)/report$",
	        releases=("v3",),
            # TODO: Remove the unstable variant after 2-3 releases
            # https://github.com/element-hq/synapse/issues/17373
	        unstable=hs.config.experimental.msc4151_enabled,
            v1=False,
	    )
```

File: synapse/rest/client/reporting.py [hunk 2]
```diff
@@ -155,8 +157,49 @@ async def on_POST(
         return 200, {}
 
 
+class ReportRoomRestServlet(RestServlet):
+    """This endpoint lets clients report a room for abuse.
+
+    Introduced by MSC4151: https://github.com/matrix-org/matrix-spec-proposals/pull/4151
+    """
+
+    PATTERNS = client_patterns("/rooms/(?P<room_id>[^/]*)/report$")
```
Comment (anoadragon453) [inline synapse/rest/client/reporting.py:166]: Note that the default value for the `releases` argument is `(r0, v3)`. I assume we only want to make this available on `v3`?

Note the unstable endpoint was available on `r0` as well.

Comment (turt2live) [reply synapse/rest/client/reporting.py:166]: For cleanliness it should be v3 only - is there a common way to declare that? It's not great that the endpoint was on r0 all this time too :(

File: synapse/rest/client/reporting.py (hunk 1 above)
Comment (turt2live) [reply synapse/rest/client/reporting.py:205]: actually, how would this know to use `unstable/org.matrix.msc4151` as its version?

Comment (anoadragon453) [reply synapse/rest/client/reporting.py:205]: Ah, true. In that case you could instead do:

```python
class ReportRoomRestServlet(RestServlet): 
    PATTERNS = client_patterns(
        "/rooms/(?P<room_id>[^/]*)/report$"),
        releases=("v3",),
        unstable=False,
        v1=False,
    )
 
    def __init__(self, hs: "HomeServer"):
        super().__init__()
        self.hs = hs
        self.auth = hs.get_auth()
        self.clock = hs.get_clock()
        self.store = hs.get_datastores().main

        # TODO: Remove the unstable variant after 2-3 releases
        # https://github.com/element-hq/synapse/issues/17373
        if hs.config.experimental.msc4151_enabled:
            self.PATTERNS.append(
                re.compile(
                    f"^{CLIENT_API_PREFIX}/unstable/org.matrix.msc4151"
                    "/rooms/(?P<room_id>[^/]*)/report$"
                )
            )

```

File: synapse/rest/client/reporting.py (hunk 2 above)
Comment (anoadragon453) [reply synapse/rest/client/reporting.py:166]: Not really, other than `releases=("v3",)` I'm afraid.

**Review by anoadragon453 — APPROVED** [review]
LGTM. I briefly questioned whether we should still gate the unstable endpoint behind an experimental feature flag, but it makes sense to keep that endpoint around only for those that were playing with the experimental feature.

**Pull request merged**